src/hooks/useCanaryObserver.tsx

```diff
@@ -0,0 +1,28 @@
+import { useLayoutEffect, useState } from "react";
+
+const useCanaryObserver = (canaryElement: HTMLDivElement | null) => {
+  const [isScrolled, setIsScrolled] = useState(false);
+
+  useLayoutEffect(() => {
+    if (!canaryElement) {
+      return;
+    }
+    const observer = new IntersectionObserver(
+      (mutations) => {
+        setIsScrolled(!mutations.some((mutation) => mutation.isIntersecting));
+      },
+      {
+        root: canaryElement.parentElement,
+      },
+    );
+    observer.observe(canaryElement);
+
+    return () => {
+      observer.disconnect();
+    };
+  }, [canaryElement]);
+
+  return isScrolled;
+};
+
+export default useCanaryObserver;
```

Comment:
A better name for this would be `useIsScrolled` and i'd probably move 👇🏻  inside the new hook too.

Canary was just a dumb name I came up with for the ref element that gets sent into the "scroll mine" to check if we're scrolling.

```typescript
const [canaryElement, setCanaryElement] = useState<HTMLDivElement | null>(
  null,
);
```

Then it could return:

```typescript
  const ObserveScroll = useCallback(({ children }: PropsWithChildren) => {
    return (
      <>
        <div ref={setCanaryElement} />
        {children}
      </>
    );
  }, []);

return {
  isScrolled,
  ObserveScroll,
}
```

Which can be used in the component to replace:

```typescript
<OakFlex
  $flexGrow={1}
  $flexDirection="column"
  $overflow="auto"
  $bt="border-solid-s"
  $borderColor={isScrolled ? "border-neutral-lighter" : "transparent"}
>
  <div ref={setCanaryElement} />
  {children}
</OakFlex>
```

With:

```typescript
<OakFlex
  $flexGrow={1}
  $flexDirection="column"
  $overflow="auto"
  $bt="border-solid-s"
  $borderColor={isScrolled ? "border-neutral-lighter" : "transparent"}
>
  <ObserveScroll>
    {children}
  </ObserveScroll>
</OakFlex>
```

👇🏻 this is the whole thing put together

```suggestion
import React, {
  PropsWithChildren,
  useCallback,
  useLayoutEffect,
  useState,
} from "react";

const useIsScrolled = () => {
  const [isScrolled, setIsScrolled] = useState(false);
  const [canaryElement, setCanaryElement] = useState<HTMLDivElement | null>(
    null,
  );
  const ObserveScroll = useCallback(({ children }: PropsWithChildren) => {
    return (
      <>
        <div ref={setCanaryElement} />
        {children}
      </>
    );
  }, []);

  useLayoutEffect(() => {
    if (!canaryElement) {
      return;
    }
    const observer = new IntersectionObserver(
      (mutations) => {
        setIsScrolled(!mutations.some((mutation) => mutation.isIntersecting));
      },
      {
        root: canaryElement.parentElement,
      },
    );
    observer.observe(canaryElement);

    return () => {
      observer.disconnect();
    };
  }, [canaryElement]);

  return { isScrolled, ObserveScroll };
};

export default useIsScrolled;
```